python_modules/libraries/dagster-tableau/dagster_tableau/components/tableau_component.py

```diff
@@ -273,6 +282,96 @@ async def write_state_to_path(self, state_path: Path) -> None:
         # Serialize and write to path
         state_path.write_text(dg.serialize_value(workspace_data))
 
+    def execute_embedded_data_sources(
+        self, context: dg.AssetExecutionContext, workspace: BaseTableauWorkspace, workbook_id: str
+    ):
+        with workspace.get_client() as client:
+            client.refresh_and_poll_workbook(workbook_id)
+            for asset_key in context.selected_asset_keys:
+                yield dg.AssetMaterialization(
+                    asset_key=asset_key,
+                )
+
+    def execute_published_data_sources(
+        self, context: dg.AssetExecutionContext, workspace: BaseTableauWorkspace
+    ):
+        specs_by_data_source_id = {
+            TableauDataSourceMetadataSet.extract(spec.metadata).id: spec
+            for spec in context.assets_def.specs
+        }
+        assert all(specs_by_data_source_id.keys())
+        with workspace.get_client() as client:
+            # pyright false positive (he can't parse the 'assert' line)
+            for datasource_id in client.refresh_and_poll_data_sources(
+                specs_by_data_source_id.keys()  # pyright: ignore[reportArgumentType]
+            ):
+                yield dg.AssetMaterialization(asset_key=specs_by_data_source_id[datasource_id].key)
+
+    def build_refreshable_embedded_data_sources_asset_definition(
+        self, workbook_id: str, specs: list[dg.AssetSpec]
+    ) -> dg.AssetsDefinition:
+        @dg.multi_asset(specs=specs, can_subset=False, name=clean_name_lower(workbook_id))
+        def asset_fn(context: dg.AssetExecutionContext):
+            yield from self.execute_embedded_data_sources(context, self.workspace, workbook_id)
+
+        return asset_fn
+
+    def build_refreshable_published_data_sources_asset_definition(
+        self, specs: list[dg.AssetSpec]
+    ) -> dg.AssetsDefinition:
+        @dg.multi_asset(
+            specs=specs,
+            can_subset=True,
+            name=clean_name_lower(f"tableau_published_data_sources_{self.workspace.site_name}"),
```

Comment:
is site-name always a valid python identifier? e.g. could it contain a hyphen? if so, should make sure to replace any potentially problematic characters before adding it in here.